src/usr/share/torwatchdog/torwatchdog.py

```diff
@@ -15,183 +15,321 @@
+            logger.warn(message)
+
+            try:
+                mail_message = gpgmailmessage.GpgMailMessage()
+                mail_message.set_subject(config['email_subject'])
+                mail_message.set_body(message)
+                mail_message.queue_for_sending()
+            except Exception as detail:
+                logger.error('Could not send up notification. %s: %s',
+                    (type(detail).__name__, detail.message))
+                logger.error(traceback.format_exc())
         prior_status = current_status
 
-except Exception,e:
-    logger.info("Stopping tor.")
-    logger.trace(traceback.format_exc())
-    tor_process.kill()
+
+# Checks if the specified website is available over Tor.
+#
+# url: The website to check for availability.
+def is_site_up(url):
+
+    logger.debug('Checking url %s.' % url)
+
+    try:
+        urllib.urlopen(url).read()
+        logger.debug('%s is up.' % url)
+        return True
+    except Exception as detail:
+        logger.warn('Unable to reach %s. %s: %s' % (url, type(detail).__name__, detail.message))
+        logger.debug('Full reason for lookup failure: %s' % traceback.format_exc())
+        return False
+        
+    logger.trace('Done checking url %s.' % url)
+
+
+# Signal handler for SIGTERM. Kills Tor and quits when SIGTERM is received.
+#
+# signal: Object representing the signal thrown.
+# stack_frame: Represents the stack frame.
+def sig_term_handler(signal, stack_frame):
+    if tor_process != None:
+        logger.info('Stopping tor.')
+        tor_process.kill()
+    sys.exit(0)
+
+
+program_uid, program_gid = get_user_and_group_ids()
+
+config, config_helper, logger = read_configuration_and_create_logger(program_uid, program_gid)
+
+tor_process = None
+try:
+    # Read gpgmailer watch directory from the gpgmailer config file
+    gpgmailmessage.GpgMailMessage.configure()
+
+    # Non-root users cannot create files in /run, so create a directory that can be written to.
+    #   Full access to user only.
+    create_directory(system_pid_dir, program_pid_dirs, program_uid, program_gid,
+        stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+
+    # Make the Tor data directory.
+    #   Full access to user only.
+    create_directory(system_data_dir, tor_data_dirs, program_uid, program_gid,
+        stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+     
+    # Configuration has been read and directories setup. Now drop permissions forever.
+    drop_permissions_forever(program_uid, program_gid)
+
+    configure_tor_proxy(config)
+
+    daemon_context = setup_daemon_context(config_helper.get_log_file_handle(), program_uid, program_gid)
+
+    tor_process = start_tor(config)
+
+    with daemon_context:
+        main_loop(config)
+ 
+except Exception as e:
+    logger.critical('Fatal %s: %s\n' % (type(e).__name__, e.message))
+    logger.critical(traceback.format_exc())
+    if tor_process != None:
```

Comment:
What am I going to put for the stack frame parameter? It doesn't quite fit. Also, I don't think 3 lines is justification enough to break out into its own method.